target/cheri-common/cheri_tagmem.c

```diff
@@ -519,117 +511,149 @@ void cheri_tag_set(CPUArchState *env, target_ulong vaddr, int reg, hwaddr* ret_p
 }
 
 bool cheri_tag_get(CPUArchState *env, target_ulong vaddr, int reg,
                   hwaddr *ret_paddr, int *prot, uintptr_t pc)
 {
-    uint64_t tag;
-    CheriTagBlock *tagblk = cheri_tag_get_block(
-        env, vaddr, MMU_DATA_CAP_LOAD, reg, 0, pc, ret_paddr, &tag, prot);
-    bool result = tagblock_get_tag(tagblk, CAP_TAGBLK_IDX(tag));
+
+    const int mmu_idx = cpu_mmu_index(env, false);
+    void *host_addr = probe_read(env, vaddr, 1, mmu_idx, pc);
+    handle_paddr_return(read);
+
+    uintptr_t tagmem_flags;
+    void *tagmem =
+        get_tagmem_from_iotlb_entry(env, vaddr, mmu_idx, false, &tagmem_flags);
+
+    if (prot) {
+        *prot = 0;
+        if (tagmem_flags & TLBENTRYCAP_FLAG_CLEAR)
+            *prot |= PAGE_LC_CLEAR;
+        if (tagmem_flags & TLBENTRYCAP_FLAG_TRAP)
+            *prot |= PAGE_LC_TRAP;
+    }
+
+    // Squash happens in the caller, so read the tagblk even if
+    // TLBENTRYCAP_FLAG_CLEAR
+    bool result =
+        (tagmem == ALL_ZERO_TAGBLK)
+            ? 0
+            : tagblock_get_tag_tagmem(tagmem, page_vaddr_to_tag_offset(vaddr));
+
     // XXX: Not atomic w.r.t. writes to tag memory
-    qemu_maybe_log_instr_extra(env, "    Cap Tag Read [" TARGET_FMT_lx "/"
-        RAM_ADDR_FMT "] -> %d\n", vaddr, (ram_addr_t)(tag * CHERI_CAP_SIZE), result);
+    qemu_maybe_log_instr_extra(
+        env, "    Cap Tag Read [" TARGET_FMT_lx "/" RAM_ADDR_FMT "] -> %d\n",
+        vaddr, qemu_ram_addr_from_host(host_addr), result);
     return result;
 }
 
-/*
- * QEMU currently tells the kernel that there are no caches installed
- * (xref target/mips/translate_init.inc.c MIPS_CONFIG1 definition)
- * so we're kind of free to make up a line size here.  For simplicity,
- * we pretend that our cache lines always contain 8 capabilities.
- */
-#define CAP_TAG_GET_MANY_SHFT    3
 int cheri_tag_get_many(CPUArchState *env, target_ulong vaddr, int reg,
         hwaddr *ret_paddr, uintptr_t pc)
 {
-    uint64_t tag;
-    int prot;
-    CheriTagBlock *tagblk =
-        cheri_tag_get_block(env, vaddr, MMU_DATA_CAP_LOAD, reg,
-                            CAP_TAG_GET_MANY_SHFT, pc, ret_paddr, &tag, &prot);
+
+    const int mmu_idx = cpu_mmu_index(env, false);
+    probe_read(env, vaddr, 1, mmu_idx, pc);
+    handle_paddr_return(read);
+
+    uintptr_t tagmem_flags;
+    void *tagmem =
+        get_tagmem_from_iotlb_entry(env, vaddr, mmu_idx, false, &tagmem_flags);
+
+    int result =
+        ((tagmem == ALL_ZERO_TAGBLK) || (tagmem_flags & TLBENTRYCAP_FLAG_CLEAR))
+            ? 0
+            : tagblock_get_tag_many_tagmem(tagmem,
+                                           page_vaddr_to_tag_offset(vaddr));
 
     /*
-     * XXX Right now, the sole consumer of this function is CLoadTags, and
-     * we let it "see around" the TLB capability load inhibit.  That should
-     * perhaps change?
+     * The Mips CLoadTags, can "see around" the TLB capability load inhibit.
+     * That should perhaps change?
```

Comment:
Is that true only of MIPS?  I don't think there's harm in making CLoadTags respect the tag-clearing or tag-load-trapping behaviors.  It's used very rarely, and at present exclusively by the kernel through the dmap region which never has those attributes set (that is, the dmap always permits cap loads and stores).